pydantic_ai_slim/pydantic_ai/toolsets/filtered.py

```diff
@@ -0,0 +1,24 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import Callable
+
+from .._run_context import AgentDepsT, RunContext
+from ..tools import ToolDefinition
+from . import AbstractToolset
+from .individually_prepared import IndividuallyPreparedToolset
+
+
+@dataclass(init=False)
+class FilteredToolset(IndividuallyPreparedToolset[AgentDepsT]):
+    """A toolset that filters the tools it contains using a filter function."""
+
+    def __init__(
+        self,
+        toolset: AbstractToolset[AgentDepsT],
+        filter_func: Callable[[RunContext[AgentDepsT], ToolDefinition], bool],
```

Comment:
```suggestion
        filter_func: Callable[[RunContext[AgentDepsT], list[ToolDefinition]], list[ToolDefinition]],
```
? Or `filter_func: Callable[[RunContext[AgentDepsT], list[ToolDefinition]], list[bool]]`?

It feels like this might be more practical in some cases, e.g., if you wanted to select the 10 most relevant tools (using some information in deps as a way to influence the selector function). I'm fine keeping as is if you think it's better to have it be a function that operates on individual items.